zetaclient/chains/solana/observer/inbound.go

```diff
@@ -0,0 +1,360 @@
+	clienttypes "github.com/zeta-chain/zetacore/zetaclient/types"
+	"github.com/zeta-chain/zetacore/zetaclient/zetacore"
+)
+
+const (
+	// MaxSignaturesPerTicker is the maximum number of signatures to process on a ticker
+	MaxSignaturesPerTicker = 100
+)
+
+// WatchInbound watches Solana chain for inbounds on a ticker.
+func (ob *Observer) WatchInbound(ctx context.Context) error {
+	app, err := zctx.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	ticker, err := clienttypes.NewDynamicTicker(
+		fmt.Sprintf("Solana_WatchInbound_%d", ob.Chain().ChainId),
+		ob.GetChainParams().InboundTicker,
+	)
+	if err != nil {
+		ob.Logger().Inbound.Error().Err(err).Msg("error creating ticker")
+		return err
+	}
+	defer ticker.Stop()
+
+	ob.Logger().Inbound.Info().Msgf("WatchInbound started for chain %d", ob.Chain().ChainId)
+	sampledLogger := ob.Logger().Inbound.Sample(&zerolog.BasicSampler{N: 10})
+
+	for {
+		select {
+		case <-ticker.C():
+			if !app.IsInboundObservationEnabled(ob.GetChainParams()) {
+				sampledLogger.Info().
+					Msgf("WatchInbound: inbound observation is disabled for chain %d", ob.Chain().ChainId)
+				continue
+			}
+			err := ob.ObserveInbound(ctx)
+			if err != nil {
+				ob.Logger().Inbound.Err(err).Msg("WatchInbound: observeInbound error")
+			}
+		case <-ob.StopChannel():
+			ob.Logger().Inbound.Info().Msgf("WatchInbound stopped for chain %d", ob.Chain().ChainId)
+			return nil
+		}
+	}
+}
+
+// ObserveInbound observes the Bitcoin chain for inbounds and post votes to zetacore.
+func (ob *Observer) ObserveInbound(ctx context.Context) error {
+	chainID := ob.Chain().ChainId
+	pageLimit := solanarpc.DefaultPageLimit
+
+	// scan from gateway 1st signature if last scanned tx is absent in the database
+	// the 1st gateway signature is typically the program initialization
+	if ob.LastTxScanned() == "" {
+		lastSig, err := solanarpc.GetFirstSignatureForAddress(ctx, ob.solClient, ob.gatewayID, pageLimit)
+		if err != nil {
+			return errors.Wrapf(err, "error GetFirstSignatureForAddress for chain %d address %s", chainID, ob.gatewayID)
+		}
+		ob.WithLastTxScanned(lastSig.String())
+	}
+
+	// get all signatures for the gateway address since last scanned signature
+	lastSig := solana.MustSignatureFromBase58(ob.LastTxScanned())
+	signatures, err := solanarpc.GetSignaturesForAddressUntil(ctx, ob.solClient, ob.gatewayID, lastSig, pageLimit)
+	if err != nil {
+		ob.Logger().Inbound.Err(err).Msg("error GetSignaturesForAddressUntil")
+		return err
+	}
+	if len(signatures) > 0 {
+		ob.Logger().Inbound.Info().Msgf("ObserveInbound: got %d signatures for chain %d", len(signatures), chainID)
+	}
+
+	// loop signature from oldest to latest to filter inbound events
+	for i := len(signatures) - 1; i >= 0; i-- {
+		sig := signatures[i]
+		sigString := sig.Signature.String()
+
+		// process successfully signature only
+		if sig.Err == nil {
+			txResult, err := ob.solClient.GetTransaction(context.TODO(), sig.Signature, &rpc.GetTransactionOpts{})
```

Comment:
Can we use the Context for the observer?

Reply:
https://github.com/zeta-chain/node/pull/2465/commits/919260b78fba3d5a2bfdea2c45fb9c30e20fb875